# Fragrance Segmentation – Unsupervised Clustering Analysis

_Course: Machine Learning_

- Name: Muhammad Umar Uz Zaman  
- Student ID: STU1197819  
- Goal: I explore a large fragrance dataset to understand:
  - How perfumes naturally group together based on their scent notes and key characteristics
  - Which fragrance segments tend to be more affordable versus more premium or luxury
  - How these segments can support targeted marketing, product positioning, and personalised recommendations
- Dataset: `Fragrance Dataset 4037.csv` (records of perfumes with information on notes and other descriptive attributes)

**Phases of this notebook**

Intro & setup  
Phase 1 – Data loading & structure check  
Phase 2 – Data cleaning, preparation & feature selection  
Phase 3 – Core EDA (distributions, relationships & feature importance)  
Phase 4 – Clustering models (K-Means, Hierarchical, DBSCAN) & evaluation  
Phase 5 – Cluster profiling & business insights  
Phase 6 – Wrap-up, limitations & export for management report


### Phase 1

In [9]:
# Core libraries
import numpy as np
import pandas as pd

# Visualisation
import matplotlib.pyplot as plt
import seaborn as sns

# Clustering and preprocessing
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.cluster import KMeans, AgglomerativeClustering, DBSCAN
from sklearn.metrics import silhouette_score, davies_bouldin_score

# Hierarchical clustering tools for dendrograms
from scipy.cluster.hierarchy import dendrogram, linkage

# Plotting style
sns.set(style="whitegrid")
plt.rcParams["figure.figsize"] = (8, 5)
plt.rcParams["axes.titlesize"] = 12
plt.rcParams["axes.labelsize"] = 11
plt.rcParams["figure.dpi"] = 100


In [13]:
# Path to the fragrance dataset
data_path = "Fragrance Dataset.csv"  # Use the CSV file instead of Excel

# Load the dataset
df = pd.read_csv(data_path)

print("Shape of the dataset (rows, columns):", df.shape)
print("\nColumn names:")
print(df.columns.tolist())

# First few rows to see how the raw data looks
df.head()


Shape of the dataset (rows, columns): (1000, 10)

Column names:
['brand', 'title', 'type', 'price', 'priceWithCurrency', 'available', 'availableText', 'sold', 'lastUpdated', 'itemLocation']


brand                                              title  \
0            Dior  Christian Dior Sauvage Men's EDP 3.4 oz Fragra...   
1         AS SHOW  A-v-entus Eau de Parfum 3.3 oz 100ML Millesime...   
2       Unbranded                   HOGO BOSS cologne For Men 3.4 oz   
3  Giorgio Armani  Acqua Di Gio by Giorgio Armani 6.7 Fl oz Eau D...   
4         Lattafa  Lattafa Men's Hayaati Al Maleky EDP Spray 3.4 ...   

              type   price priceWithCurrency  available  \
0    Eau de Parfum   84.99      US $84.99/ea       10.0   
1    Eau de Parfum  109.99        US $109.99        8.0   
2  Eau de Toilette  100.00        US $100.00       10.0   
3  Eau de Toilette     NaN      US $44.99/ea        2.0   
4       Fragrances   16.91         US $16.91        NaN   

                           availableText   sold                lastUpdated  \
0      More than 10 available / 116 sold  116.0  May 24, 2024 10:03:04 PDT   
1                  8 available / 48 sold   48.0  May 23, 2024 23:07:49 PDT   
2       More than 10 available / 27 sold   27.0  May 22, 2024 21:55:43 PDT   
3                 2 available / 159 sold  159.0  May 24, 2024 03:30:43 PDT   
4  Limited quantity available / 156 sold  156.0  May 24, 2024 07:56:25 PDT   

                             itemLocation  
0     Allen Park, Michigan, United States  
1                Atlanta, Georgia, Canada  
2       Dearborn, Michigan, United States  
3  Reinholds, Pennsylvania, United States  
4       Brooklyn, New York, United States

In [14]:
# Overall information about the dataframe
df.info()

# Missing values per column
print("\nMissing values per column:")
print(df.isna().sum().sort_values(ascending=False))

# Summary statistics for numeric columns
numeric_summary = df.describe(include=[np.number]).T
numeric_summary


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   brand              999 non-null    object 
 1   title              1000 non-null   object 
 2   type               997 non-null    object 
 3   price              996 non-null    float64
 4   priceWithCurrency  1000 non-null   object 
 5   available          889 non-null    float64
 6   availableText      997 non-null    object 
 7   sold               994 non-null    float64
 8   lastUpdated        947 non-null    object 
 9   itemLocation       1000 non-null   object 
dtypes: float64(3), object(7)
memory usage: 78.2+ KB

Missing values per column:
available            111
lastUpdated           53
sold                   6
price                  4
availableText          3
type                   3
brand                  1
title                  0
priceWithCurrency      0
itemLocation

count        mean          std  min    25%    50%    75%       max
price      996.0   46.497279    35.590865  3.0  22.99  35.71   59.0    259.09
available  889.0   20.046119    61.547985  2.0   5.00  10.00   10.0    842.00
sold       994.0  766.266600  3200.971733  1.0  14.00  49.50  320.5  54052.00

From this structural check I can see that the dataset has a mix of numeric and categorical information:

- There are three numeric columns: `price`, `available` and `sold`. These capture the commercial side of each perfume listing: how much it costs, how many units are available and how many units have been sold.
- The remaining columns are stored as text, including `brand`, `title`, `type`, `priceWithCurrency`, `availableText`, `lastUpdated` and `itemLocation`. These fields describe the product, the listing text, the time of the last update and where the item is located.

In terms of data quality:

- The `price` and `sold` columns are almost complete, with only a handful of missing values, so I expect to rely on them strongly in my analysis.
- The `available` column has more missing entries, which suggests that availability information is not always reported or not always parsed correctly.
- The `lastUpdated` and `availableText` fields are mostly complete but are stored as free text, which means they will be harder to use directly in a clustering algorithm.
- The basic statistics show that `price`, `available` and especially `sold` are all quite skewed, with a few very large values. This is typical for online marketplace data, where a small number of listings can have very high stock or sales.

Overall, this first check reassures me that I have several usable numeric features for clustering, while also highlighting that I need to be careful with skewed distributions and with how I handle missing values in `available`. Before I move on to cleaning and feature selection, I briefly look at the structure of the most important categorical columns.


In [15]:
# Quick look at the structure of key categorical columns
cat_cols = ["brand", "type", "itemLocation"]

for col in cat_cols:
    print(f"\nColumn: {col}")
    print(f"Number of unique values: {df[col].nunique()}")
    print("Most frequent values:")
    print(df[col].value_counts().head(10))



Column: brand
Number of unique values: 246
Most frequent values:
brand
Giorgio Armani        60
Paco Rabanne          45
Ralph Lauren          41
Unbranded             39
Armaf                 34
Dolce&Gabbana         32
Yves Saint Laurent    30
Versace               20
Calvin Klein          17
Cologne               16
Name: count, dtype: int64

Column: type
Number of unique values: 63
Most frequent values:
type
Eau de Toilette    444
Eau de Parfum      285
Eau de Cologne      56
Perfume             50
Fragrances          32
Eau De Parfum       15
Parfum              13
Eau de Perfume      10
Eau De Toilette      9
Gift Sets            6
Name: count, dtype: int64

Column: itemLocation
Number of unique values: 276
Most frequent values:
itemLocation
Dallas, Texas, United States             126
Hackensack, New Jersey, United States    105
Brooklyn, New York, United States         76
Miami, Florida, United States             41
Edison, New Jersey, United States         35
Houston, Texas, 

The categorical overview gives me a clearer picture of the breadth of the dataset:

- The `brand` column has 246 unique values, with a long tail beyond the most common brands such as Giorgio Armani, Paco Rabanne and Ralph Lauren. This tells me that brand is highly granular and better suited for describing clusters later rather than for direct use as a modelling feature.
- The `type` column has 63 unique values, but the majority of listings fall into a few familiar fragrance formats such as Eau de Toilette, Eau de Parfum, Eau de Cologne and Perfume. This suggests that I can simplify `type` into a smaller set of meaningful groups without losing much information.
- The `itemLocation` field contains many different city and state combinations. This provides useful commercial context (for example, where sellers are based), but the variety of locations means that it is unlikely to be a stable driver of fragrance similarity. I plan to use it mainly for interpretation rather than as a clustering feature.

These observations already point towards a feature-selection strategy where I focus on the core numeric variables (`price`, `available`, `sold`) and a simplified version of `type` for clustering, while keeping `brand` and `itemLocation` for later profiling of the resulting segments.


### Phase 2 – Data cleaning and feature selection

In this phase I prepare the dataset for clustering by cleaning key variables and constructing a first set of candidate features.

My main objectives are:

- To handle missing values in a transparent way, especially for the commercial variables `price`, `available` and `sold`.
- To simplify the `type` column into a smaller number of fragrance-format groups that are easier to interpret.
- To engineer a small set of numeric features that capture price level and sales behaviour in a way that is suitable for clustering algorithms.

For modelling I want to focus on variables that are both commercially meaningful and reasonably complete. I therefore treat `price` and `sold` as essential features and I avoid discarding too many rows because of missing `available`. The high-cardinality columns `brand` and `itemLocation` are retained for later interpretation of clusters rather than being turned directly into model features.


In [16]:
# Create a working copy for cleaning and feature engineering
df_clean = df.copy()

# Drop rows with missing price or sold (key commercial signals, only a few rows missing)
rows_before = df_clean.shape[0]
df_clean = df_clean.dropna(subset=["price", "sold"])
rows_after = df_clean.shape[0]

print(f"Rows before dropping: {rows_before}")
print(f"Rows after dropping rows with missing price or sold: {rows_after}")

# Impute 'available' with the median of the observed values
available_median = df_clean["available"].median()
df_clean["available"] = df_clean["available"].fillna(available_median)

print(f"\nMedian value used to fill 'available': {available_median}")

# Check remaining missing values
print("\nRemaining missing values per column:")
print(df_clean.isna().sum())


Rows before dropping: 1000
Rows after dropping rows with missing price or sold: 990

Median value used to fill 'available': 10.0

Remaining missing values per column:
brand                 1
title                 0
type                  3
price                 0
priceWithCurrency     0
available             0
availableText         0
sold                  0
lastUpdated          47
itemLocation          0
dtype: int64


In [17]:
# Simplify the 'type' column into a smaller set of fragrance-format groups
def simplify_type(value):
    if isinstance(value, str):
        text = value.lower()
        if "toilette" in text:
            return "Eau de Toilette"
        if "parfum" in text:
            return "Eau de Parfum"
        if "cologne" in text:
            return "Eau de Cologne"
        if "perfume" in text:
            return "Perfume"
    return "Other"

df_clean["type_group"] = df_clean["type"].apply(simplify_type)

# Check the distribution of the simplified groups
df_clean["type_group"].value_counts()


type_group
Eau de Toilette    454
Eau de Parfum      328
Other               79
Eau de Cologne      68
Perfume             61
Name: count, dtype: int64

In [18]:
# Engineer additional numeric features to capture scale and sales intensity
df_clean["log_price"] = np.log1p(df_clean["price"])
df_clean["log_sold"] = np.log1p(df_clean["sold"])
df_clean["sold_per_available"] = df_clean["sold"] / (df_clean["available"] + 1)

# Quick summary of the key numeric features
df_clean[["price", "available", "sold",
          "log_price", "log_sold", "sold_per_available"]].describe().T


count        mean          std       min        25%  \
price               990.0   46.184202    34.843593  3.000000  22.990000   
available           990.0   19.036364    58.396700  2.000000   5.000000   
sold                990.0  759.216162  3196.038085  1.000000  14.000000   
log_price           990.0    3.627914     0.670125  1.386294   3.177637   
log_sold            990.0    4.283952     2.062323  0.693147   2.708050   
sold_per_available  990.0   63.139165   296.270648  0.020000   1.782670   

                          50%         75%           max  
price               35.710000   58.997500    259.090000  
available           10.000000   10.000000    842.000000  
sold                48.000000  315.250000  54052.000000  
log_price            3.603049    4.094303      5.561028  
log_sold             3.891820    5.756525     10.897720  
sold_per_available   5.842424   21.738636   4913.818182

The cleaning steps reduce the dataset from 1,000 to 990 rows by removing listings with missing values in `price` or `sold`. This is less than 1% of the data, so I regard the impact on the overall distribution as minimal while gaining the benefit of having complete information for the core commercial variables.

For `available` I fill missing values with the median of 10 units. This choice keeps the distribution stable because:
- The median is robust to extreme high-stock listings.
- A value of 10 fits well with the observed quartiles and typical stock levels in the dataset.

A small number of missing values remain in `brand`, `type` and `lastUpdated`, but these variables are mainly used for interpretation and descriptive context rather than as direct clustering features. I therefore keep the affected rows and treat the missing entries as minor limitations that I will mention later in the written report.

The new engineered features behave as expected:

- `log_price` compresses the original price range from 3 to around 259 into a narrower scale, which helps reduce the influence of very expensive listings while still distinguishing between low-, mid- and high-price perfumes.
- `log_sold` performs a similar compression for the `sold` column, where a few listings show extremely high sales counts compared with the majority of perfumes.
- `sold_per_available` captures an intensity measure that approximates demand relative to stock. The distribution is still skewed, but it provides a useful signal for identifying listings that sell quickly compared with their available quantity.

Together, these features give me a compact numerical description of each listing that combines price level, scale of sales and sales efficiency. In the next step I add a log-transformed version of `available` and then summarise the final set of features that I intend to use for clustering.


In [19]:
# Add a log-transformed version of 'available'
df_clean["log_available"] = np.log1p(df_clean["available"])

# Candidate numeric features for clustering
candidate_features = ["log_price", "log_sold", "log_available", "sold_per_available"]

print("Candidate features for clustering:")
print(candidate_features)

# Quick summary of these features
df_clean[candidate_features].describe().T


Candidate features for clustering:
['log_price', 'log_sold', 'log_available', 'sold_per_available']


count       mean         std       min       25%  \
log_price           990.0   3.627914    0.670125  1.386294  3.177637   
log_sold            990.0   4.283952    2.062323  0.693147  2.708050   
log_available       990.0   2.300874    0.851423  1.098612  1.791759   
sold_per_available  990.0  63.139165  296.270648  0.020000  1.782670   

                         50%        75%          max  
log_price           3.603049   4.094303     5.561028  
log_sold            3.891820   5.756525    10.897720  
log_available       2.397895   2.397895     6.736967  
sold_per_available  5.842424  21.738636  4913.818182

In [20]:
# Tidy remaining missing values in descriptive (non-feature) columns
df_clean["brand"] = df_clean["brand"].fillna("Unknown brand")
df_clean["lastUpdated"] = df_clean["lastUpdated"].fillna("Unknown date")

# type_group should already be filled by the simplify_type function,
# but this is a safety check in case any missing values remain.
df_clean["type_group"] = df_clean["type_group"].fillna("Other")

print("Remaining missing values per column:")
print(df_clean.isna().sum())


Remaining missing values per column:
brand                 0
title                 0
type                  3
price                 0
priceWithCurrency     0
available             0
availableText         0
sold                  0
lastUpdated           0
itemLocation          0
type_group            0
log_price             0
log_sold              0
sold_per_available    0
log_available         0
dtype: int64


At this point I have a first version of the feature set that I plan to use as input for the clustering algorithms:

- **`log_price`** – captures the general price level of each perfume while limiting the impact of a few very expensive listings.
- **`log_sold`** – measures the scale of demand in a compressed form that keeps very high sales counts from dominating the analysis.
- **`log_available`** – reflects how much stock is advertised for a listing, again on a log scale to handle large differences between small and large sellers.
- **`sold_per_available`** – provides an intensity-style indicator that hints at how quickly products sell relative to their available quantity.

These features are deliberately numeric, compact and commercially interpretable. They describe how each perfume is positioned in terms of price, stock and sales performance, which are all important dimensions for a retailer or marketplace. I keep `brand`, `type_group` and `itemLocation` as descriptive variables that I will use later to label and interpret the resulting clusters in business language, but I do not include them directly as inputs to the clustering algorithms at this stage.

In the next phase I explore these candidate features in more depth using visualisations and correlation analysis to check whether they are informative, redundant or overly skewed for clustering.
